lib/SILOptimizer/Mandatory/DataflowDiagnostics.cpp

```diff
@@ -42,19 +42,37 @@ static void diagnoseMissingReturn(const UnreachableInst *UI,
   SILLocation FLoc = F->getLocation();
 
   Type ResTy;
+  BraceStmt *BS;
 
   if (auto *FD = FLoc.getAsASTNode<FuncDecl>()) {
     ResTy = FD->getResultInterfaceType();
+    BS = FD->getBody(/*canSynthesize=*/false);
   } else if (auto *CD = FLoc.getAsASTNode<ConstructorDecl>()) {
     ResTy = CD->getResultInterfaceType();
+    BS = FD->getBody();
   } else if (auto *CE = FLoc.getAsASTNode<ClosureExpr>()) {
     ResTy = CE->getResultType();
+    BS = CE->getBody();
   } else {
     llvm_unreachable("unhandled case in MissingReturn");
   }
 
   SILLocation L = UI->getLoc();
   assert(L && ResTy);
+  auto numElements = BS->getNumElements();
+  if (numElements > 0) {
+    auto element = BS->getElement(numElements - 1);
+    if (auto expr = element.dyn_cast<Expr *>()) {
+      if (expr->getType()->getCanonicalType() == ResTy->getCanonicalType()) {
```

Comment:
TypeBase::isEqual() is a shortcut for comparing `getCanonicalType()s`

Reply:
PR with a fix at https://github.com/apple/swift/pull/24546 .